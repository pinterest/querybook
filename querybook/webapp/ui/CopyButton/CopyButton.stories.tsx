import React from 'react';
// eslint-disable-next-line import/no-extraneous-dependencies
import centered from '@storybook/addon-centered/react';

import { CopyButton, ICopyButtonProps } from './CopyButton';

export const _CopyButton = (args: ICopyButtonProps) => <CopyButton {...args} />;
_CopyButton.args = {
    copyText: 'text to copy',
    title: 'Copy To Clipboard',
};

export default {
    title: 'Button/CopyButton',
    decorators: [centered],
};
