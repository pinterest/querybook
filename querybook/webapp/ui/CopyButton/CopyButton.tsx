import clsx from 'clsx';
import React from 'react';

import { TooltipDirection } from 'const/tooltip';
import * as Utils from 'lib/utils';
import {
    ButtonProps,
    ButtonType,
    getButtonComponentByType,
} from 'ui/Button/Button';

const DEFAULT_TOOL_TIP = 'Click To Copy';
const DEFAULT_COPIED_TOOL_TIP = 'Copied!';

export interface ICopyButtonProps extends ButtonProps {
    copyText: string | (() => string);
    icon?: string;
    title?: string;
    className?: string;

    tooltip?: string;
    copiedTooltip?: string;
    tooltipDirection?: TooltipDirection;
    type?: ButtonType;
}

export const CopyButton: React.FunctionComponent<ICopyButtonProps> = ({
    copyText,
    className = '',
    tooltip = DEFAULT_TOOL_TIP,
    copiedTooltip = DEFAULT_COPIED_TOOL_TIP,
    tooltipDirection = 'up',
    type,
    ...propsForButton
}) => {
    const [tooltipToShow, setTooltipToShow] = React.useState(tooltip);

    const Button = getButtonComponentByType(type);
    return (
        <Button
            className={clsx({
                CopyButton: true,
                [className]: className,
            })}
            aria-label={tooltipToShow}
            data-balloon-pos={tooltipDirection}
            onClick={() => {
                Utils.copy(
                    typeof copyText === 'function' ? copyText() : copyText
                );
                setTooltipToShow(copiedTooltip);
            }}
            onMouseLeave={() => setTooltipToShow(tooltip)}
            icon="copy"
            {...propsForButton}
        />
    );
};
