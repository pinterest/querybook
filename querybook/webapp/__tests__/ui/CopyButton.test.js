import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import renderer from 'react-test-renderer';

import { CopyButton } from '../../ui/CopyButton/CopyButton';

it('renders without crashing', () => {
    shallow(
        <CopyButton
            type="text"
            size="small"
            copyText="Text to be copied"
            icon="link"
            title="Test Copy"
        />
    );
});

describe('matches enzyme snapshots', () => {
    it('matches snapshot', () => {
        const wrapper = shallow(
            <CopyButton
                type="text"
                size="small"
                copyText="Text to be copied"
                icon="link"
                title="Test Copy"
            />
        );
        const serialized = toJson(wrapper);
        expect(serialized).toMatchSnapshot();
    });
});

describe('matches test renderer snapshot', () => {
    it('serializes the styles', () => {
        const output = renderer.create(
            <CopyButton
                type="text"
                size="small"
                copyText="Text to be copied"
                icon="link"
                title="Test Copy"
            />
        );
        expect(output).toMatchSnapshot();
    });
});
