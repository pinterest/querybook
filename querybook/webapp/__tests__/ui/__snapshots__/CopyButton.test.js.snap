// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches enzyme snapshots matches snapshot 1`] = `
<Button
  aria-label="Click To Copy"
  className="CopyButton"
  data-balloon-pos="up"
  icon="link"
  onClick={[Function]}
  onMouseLeave={[Function]}
  size="small"
  title="Test Copy"
/>
`;

exports[`matches test renderer snapshot serializes the styles 1`] = `
<span
  aria-label="Click To Copy"
  className="StyledButton-aqwbwz-0 jBXKFp Button CopyButton flex-row"
  color="var(--text-color)"
  data-balloon-pos="up"
  fontWeight="700"
  onClick={[Function]}
  onMouseLeave={[Function]}
  size="small"
>
  <span
    className=" Icon "
    dangerouslySetInnerHTML={
      Object {
        "__html": "<svg xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"24\\" height=\\"24\\" viewBox=\\"0 0 24 24\\" fill=\\"none\\" stroke=\\"currentColor\\" stroke-width=\\"2\\" stroke-linecap=\\"round\\" stroke-linejoin=\\"round\\" class=\\"feather feather-link\\"><path d=\\"M10 13a5 5 0 0 0 7.54.54l3-3a5 5 0 0 0-7.07-7.07l-1.72 1.71\\"></path><path d=\\"M14 11a5 5 0 0 0-7.54-.54l-3 3a5 5 0 0 0 7.07 7.07l1.71-1.71\\"></path></svg>",
      }
    }
  />
  <span>
    Test Copy
  </span>
</span>
`;
